### DISCUSSION TITLE
esp proposal as part of add_sa

### DISCUSSION THREAD

**Discussion by musaleh17** [strongswan/strongswan#1543] [General]
Hi,

in swanctl conf file we define esp_proposals for example aes128gcm128-x25519, I expected to receive the proposal as part add_sa.
is it right? I thought that the integrity algorithm variable in add_sa kernel struct should store the algorithm that was inserted in the conf file!

**Comment by tobiasbrunner** [comment]
The algorithms are passed in the `enc_alg` and `int_alg` members of the `kernel_ipsec_add_sa_t` struct that's passed to `kernel_ipsec_t::add_sa()`. But with `aes128gcm128` there won't be an integrity algorithm, only `enc_alg` will be set (as is the case with all AEAD algorithms).

**Reply by musaleh17** [reply]
Is this because GCM is providing also integrity?
specifically, what I need is to know what is the size of the encryption key? 128 or 256? is there a way to know through the kernel_ipsec_add_sa_t?

**Reply by tobiasbrunner** [reply]
> Is this because GCM is providing also integrity?

Yes.

> specifically, what I need is to know what is the size of the encryption key? 128 or 256? is there a way to know through the kernel_ipsec_add_sa_t?

The key and its size is in the `enc_key` member of the mentioned struct. Note that this includes additional key material for the salt, which is 32 bit long for AES-GCM. So for `aes128gcm128`, `enc_key.len` will be 20, for `aes256gcm128` if would be 36.

**Reply by musaleh17** [reply]
Thanks Tobias!

last question, do the 32 bits for salt come at the start of enc_key.ptr or at the end? 
is GCM IV part of kernel_ipsec_add_sa_t?

**Reply by tobiasbrunner** [reply]
> last question, do the 32 bits for salt come at the start of enc_key.ptr or at the end?

At the end, see [RFC 4106, section 8.1](https://datatracker.ietf.org/doc/html/rfc4106#section-8.1).

> is GCM IV part of kernel_ipsec_add_sa_t?

IVs are allocated per ESP packet, so no. (For GCM mode a counter is usually used to generate IV values.)

**Reply by musaleh17** [reply]
I see,
thanks 👍
